Month 13/step 1

Shiker2032 opened the pull request.

Shiker2032 committed:
Тестирование circle и button

Shiker2032 committed:
Тестирование разворота строки

Shiker2032 committed:
доработка тестирования ui и разворота строки

Shiker2032 committed:
Тестирование сортировки выбором и пузырьком

Shiker2032 committed:
2e2 тестирование страницы разворота строки

Shiker2032 committed:
e2e тестирование страницы Фибоначчи

Shiker2032 committed:
e2e тестирование страницы стека

Shiker2032 committed:
2e2 тестирование страницы очереди

Shiker2032 committed:
2у2 тестирование страницы связного списка

Shiker2032 committed:
Подготовил к проверке

Shiker2032 committed:
Удалил сэмпл тесты

gennady-bars commented on cypress.config.ts:
```diff
@@ -0,0 +1,7 @@
+
+export default defineConfig({
+  e2e: {
+    baseUrl: "http://localhost:3000",
```
Отлично указан базовый урл в конфиге

gennady-bars commented on src/components/sorting-page/sorting-page.tsx:
```diff
@@ -17,7 +19,15 @@ export const SortingPage: React.FC = () => {
   useEffect(() => {
-    setRenderArray([...prepareArray(1, 100)])
+    if (initialArray) {
+      const preparedArray: any = []
```
`const preparedArray: any` нужно типизировать правильно массив

gennady-bars commented on cypress/support/e2e.ts:
```diff
@@ -0,0 +1,20 @@
+import './commands'
+
+// Alternatively you can use CommonJS syntax:
+// require('./commands')
```
**Можно лучше**  (какие-то уже у Вас выполнены, возможно)

* Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.
* В `devDependencies` указывается то, что необходимо для сборки и разработки приложения, а в `dependencies` нужно указывать то, что необходимо непосредственно для работы приложения (express, uuid). Все линтеры, тесты и тд нужно только в `devDependencies` указывать, чтобы зря не раздувать кол-во пакетов в собранном для деплоя приложении.
* В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log
* хорошей практикой является подключение `normalize.css`, чтобы **унифицировать** стилизацию под разные браузеры. Подробнее тут https://htmlacademy.ru/blog/html/about-normalize-css . И эти стили нужно получать **первее**всех других, чтобы переопределить изначально стили по умолчанию.
* Лучше не обращаться к **нативному** тегу, а задать для него **отдельный** css-класс.
При изменении **тэгов** в шаблоне в таком случае код будет продолжать работать, если в стили прописаны через имена классов
[Хорошая статья](https://frontstuff.io/you-need-to-stop-targeting-tags-in-css) не эту тему.
* все **цвета** лучше вынести в `css`-переменные и использовать их, чтобы можно было мгновенно понимать, что это за цвет, и можно было бы подкорректировать его в одном месте  и не искать дублирования во всем коде

* Чтобы сделать сайт более доступным для пользователя, нужно стараться придерживаться **семантической верстки**.
Несколько полезных статей на эту тему:
•	[Что такое семантическая вёрстка и зачем она нужна](https://htmlacademy.ru/blog/html/semantics)
•	[Семантика для циников](https://www.youtube.com/watch?v=B5pomLr059U)

gennady-bars commented on cypress/e2e/tests/list.cy.js:
```diff
@@ -0,0 +1,163 @@
+
+describe("List page tests", () => {
+  const initialArrayRef = [0, 34, 8, 1]
+  const changingColor = "rgb(210, 82, 225)"
+  const defaultColor = "rgb(0, 50, 255)"
+  const modifiedColor = "rgb(127, 224, 81)"
```
```js

  const changingColor = "rgb(210, 82, 225)"
  const defaultColor = "rgb(0, 50, 255)"
  const modifiedColor = "rgb(127, 224, 81)"

```

эти константы в каждом файле. Лучше вынести в отдельный файл и импортировать в каждый тест

gennady-bars commented on src/components/sorting-page/sorting-page.test.tsx:
```diff
@@ -0,0 +1,185 @@
+
+    userEvent.click(selectionRadio)
+    userEvent.click(ascBtn)
+    await waitFor(() => expect(screen.getByTestId("column-div").textContent).toBe("1"), {
```
**Можно лучше**

В юнит-тестах нужно было тестировать функции `bubbleSort` и `selectionSort` отдельно от самих страниц, где они используются, чтобы независимо проверять их корректность

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* все юнит-тесты проходят успешно

но есть некоторые недочеты: 

* `const preparedArray: any` нужно типизировать правильно массив
* Вам нужно добавить скрипт запуска `cypress` тестов ` "cypress": "cypress open cache clear",` в `package.json`
* cypress-тесты не проходят . При локальной сборке проекта все модульные классы `.module.css` получают уникальный **хэш**. Вот этот хэш каждый раз меняется при новой сборке проекта, поэтому у другого разработчика будут другие `css`-селекторы. В тесте нельзя использовать классы с хэшем, иначе тесты не будут проходить при следующей сборке (даже у Вас на компьютере).  Скрин https://disk.yandex.ru/i/tpVKb0VsBiwnUQ 
* Обычно 15 Snapshots в проекте делают, но  у Вас 2 шт. Возможно, где-то не покрыли тестами


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Shiker2032 committed:
Исправления после ревью

Типизировать prepared array
Добавить скрипт cypress в package
Убрать селектор с hash
вынести colors в constants
зафиксировать зависимости
поставить normalize

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
